src/status_im2/contexts/chat/messages/pin/events.cljs

```diff
@@ -33,34 +33,46 @@
-                                                  ;; Add to or remove from pinned message list, and
-                                                  ;; normalizing pinned-by property
-                                                  (let [current-message     (get already-loaded-messages
-                                                                                 message_id)
-                                                        current-message-pin (merge current-message
-                                                                                   {:pinned    pinned
-                                                                                    :pinned-by from})]
-                                                    (cond-> acc
-                                                      (nil? pinned)
-                                                      (dissoc message_id)
+(defn remove-pinned-message
+  [db pinned-message-js]
+  (update-in db
+             [:pin-messages (.-localChatId pinned-message-js)]
+             dissoc
+             (.-message_id pinned-message-js)))
 
-                                                      (and (some? pinned) (some? current-message))
-                                                      (assoc message_id current-message-pin))))
-                                                already-loaded-pin-messages
-                                                pin-messages)]
-        {:db (-> db
-                 (assoc-in [:pin-messages chat-id] all-messages)
-                 (assoc-in [:pin-message-lists chat-id]
-                           (message-list/add-many nil (vals all-messages))))}))))
+(defn add-pinned-message
+  [db pinned-message-js]
+  (let [message-js (.-pinnedMessage pinned-message-js)]
+    (if (and message-js
+             (.-message message-js))
+      (assoc-in db
+       [:pin-messages
+        (.-localChatId pinned-message-js)
+        (.-message_id pinned-message-js)]
+       (-> (.-message message-js)
+           (js->clj :keywordize-keys true)
+           data-store.messages/<-rpc
+           (assoc :pinned-by (.-pinnedBy message-js)
+                  :pinned-at (.-pinnedAt message-js))))
+      db)))
 
+(rf/defn receive-signal
+  [{:keys [db]} pin-messages-js]
+  (let [current-chat-id (db :current-chat-id)
+        db              (reduce (fn [db pin-message-js]
+                                  (let [pinned? (.-pinned pin-message-js)
+                                        chat-id (.-localChatId pin-message-js)]
+                                    (if (not= chat-id current-chat-id)
+                                      db
+                                      (if pinned?
+                                        (add-pinned-message db pin-message-js)
+                                        (remove-pinned-message db pin-message-js)))))
+                                db
+                                pin-messages-js)]
+    {:db
```

Comment:
We almost never test the business logic in event handlers, even though it's made up mostly of pure functions and it's highly recommended in the re-frame docs https://github.com/day8/re-frame/blob/976ac0fd1b1055d76775f793c8a2650f0b75051a/docs/Testing.md#what-to-test. Maybe it's a lost battle, but I'm leaving a comment anyway ;)

Reply:
definitely worth a test,t hanks :)